README.md

```diff
@@ -0,0 +1,211 @@
+
+You can convert EC private key to public key and print it:
+
+```sh
+java -jar ssl-pinning-tool.jar export -k keypair.pem -p [password]
+```
+
+### Troubleshooting
+
+Error: 
+```
+SEVERE: Failed to load private key, error: unable to read encrypted data: javax.crypto.BadPaddingException: pad block corrupted
+```
+
+This error is shown when the private key password is invalid.
+
+Alternatively, you can perform all steps above using `openssl` command.
+
+## Prepare Data Using OpenSSL
+
+In case you have your SSL certificate in a `pem` format, you are able to generate all necessary data using OpenSSL.
+
+### Generate a Key Pair
+
+This command will generate a new ECDSA key pair and store it in the `keypair.pem` file. You need to store this file securely.
+
+```sh
+openssl ecparam -name prime256v1 -genkey | openssl pkcs8 -topk8 -v2 aes-128-cbc > keypair.pem
+```
+
+**Store the key pair and private key password safely! You will need it next time you replace SSL certificate to generate new signatures.**
+
+### Prepare a Certificate Fingerprint
+
+This sequence of commands converts a server certificate stored in the `cert.pem` file to fingerprint encoded as Base64 and stored in `fingerprint.txt` file.
+
+```sh
+# Convert PEM to DER format
+openssl x509 -in cert.pem -inform PEM -outform DER -out cert.der
+
+# Compute SHA256 digest
+openssl dgst -sha256 < cert.der > fingerprint_raw.txt
+
+# Encode the digest as Base64
+openssl enc -base64 -A < fingerprint_raw.txt > fingerprint.txt
+```
+
+### Get Certificate Attributes
+
+In order to compute the signature, you need to have values of certificate common name and expiration timestamp.
+
+To obtain common name, call:
+
+```sh
+openssl x509 -noout -subject -inform der -in cert.der | sed -n '/^subject/s/^.*CN=//p'
+```
+
+To obtain expiration timestamp, call:
+
+```sh
+openssl x509 -noout -dates -inform der -in cert.der  | grep notAfter
+```
+
+Then, you need to convert the expiration timestamp to Unix epoch format (seconds after 1970/01/01).
+
+```
+TIME=`openssl x509 -noout -dates -inform der -in cert.der | grep notAfter | sed -e 's#notAfter=##'`
+
+if date --version >/dev/null 2>/dev/null; then
+  # Linux
+  date -d "$TIME" "+%s"
+else
+  # MacOSX
+  date -j -f "%b %d %H:%M:%S %Y %Z" "$TIME" "+%s"  
+fi
+```
+
+### Sign Certificate Fingerprint
+
+This sequence of commands signs the fingerprint from `fingerprint.txt` with the private key from the provided key pair file `keypair.pem` and stores the result signature as a Base64 encoded file `sign.txt`.
+
+```sh
+# Prepare a signature base string as $COMMON_NAME + '&' + $UNIXTIMESTAMP_EXPIRATION + '&' + $FINGERPRINT
+echo "$COMMON_NAME" > signature_base_string.txt
+echo "&" >> signature_base_string.txt
+echo "$UNIXTIMESTAMP_EXPIRATION" >> signature_base_string.txt
+echo "&" >> signature_base_string.txt
+cat fingerprint.txt >> signature_base_string.txt
```

Comment:
This script does not work properly, because newlines are added into the signature_base_string.txt file everytime echo is used. So the final string looks like:
```
my.domain.com
&
1543322263
&
jymEKdgGPv1zSp61CYya5c2fR9fTLe8tKnWF6857iLA=
```